src/ray/object_manager/common.cc

```diff
@@ -74,10 +74,33 @@ Status PlasmaObjectHeader::CheckHasError() const {
 
 #if defined(__APPLE__) || defined(__linux__)
 
-Status PlasmaObjectHeader::TryToAcquireSemaphore(sem_t *sem) const {
+Status PlasmaObjectHeader::TryToAcquireSemaphore(
+    sem_t *sem,
+    const std::unique_ptr<std::chrono::steady_clock::time_point> &timeout_point) const {
   // Check `has_error` first to avoid blocking forever on the semaphore.
   RAY_RETURN_NOT_OK(CheckHasError());
-  RAY_CHECK_EQ(sem_wait(sem), 0);
+  auto now = std::chrono::steady_clock::now();
+
+  if (!timeout_point) {
+    RAY_CHECK_EQ(sem_wait(sem), 0);
+  } else if (now >= *timeout_point) {
+    if (sem_trywait(sem) != 0) {
+      return Status::ChannelTimeoutError("Timed out acquiring semaphore.");
+    }
+  } else {
+    bool got_sem = false;
+    // macOS does not support sem_timedwait, so we implement a unified solution here.
+    while (std::chrono::steady_clock::now() < *timeout_point) {
+      if (sem_trywait(sem) == 0) {
+        got_sem = true;
+        break;
+      }
+    }
```

Comment:
nit: You can combine both of these cases into one by using a do...while loop

Reply:
Good call.